src/understand-tidb/the-lifecycle-of-a-statement.md

```diff
@@ -0,0 +1,64 @@
+# The lifecycle of a statement
+
+## MySQL protocol package with command and statement string
+After [connecting and getting authenticated](understand-tidb/mysql-protocol-and-session-management.md) the server is in a [statement execution loop](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/server/conn.go#L938) until the client is disconnected.
+
+The [dispatch](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/server/conn.go#L1111) function checks what kind of command was sent through the MySQL protocol and dispatches the matching function, like this snippet:
+```golang
+	switch cmd {
+...
+	case mysql.ComQuit:
+		return io.EOF
+	case mysql.ComInitDB:
+		if err := cc.useDB(ctx, dataStr); err != nil {
+			return err
+		}
+		return cc.writeOK(ctx)
+	case mysql.ComQuery: // Most frequently used command.
+		return cc.handleQuery(ctx, dataStr)
+```
+Where [mysql.ComQuery](https://github.com/pingcap/parser/blob/d4a88481405f8c59d45fc0a9c38ee24d55b9bf49/mysql/const.go#L102) is routed to [handleQuery](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/server/conn.go#L1633), which handles all different non-prepared statements (some commands like change database/schema or ping are handled directly in the dispatch function).
+
+## Statement string to Abstract Syntax Tree
+In [handleQuery](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/server/conn.go#L1633) we first needs to [parse](https://github.com/pingcap/parser/blob/d4a88481405f8c59d45fc0a9c38ee24d55b9bf49/yy_parser.go#L137) the statement string, which is done by [the parser, which is in its own repository](https://github.com/pingcap/parser) which is a MySQL compatible parser which parses statements and returns an Abstract Syntax Tree (AST) representing the statement.
+
+## AST -> Physical execution plan
+We now have a statement in AST form to handle in [handleStmt](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/server/conn.go#L1814)/[ExecuteStmt](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/session/session.go#L1620) where the Abstract Syntax Tree is [compiled](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/executor/compiler.go#L50) to a physical execution plan, including [optimizing](https://github.com/pingcap/tidb/blob/30cf15a59db11c34ffe05fc926152a43327eaa61/planner/optimize.go#L100) the execution plan, through a cost based optimizer. There are several steps in this process, such as name resolution, transaction management, [privilege checks](understand-tidb/privilege-management.md), handling given hints, etc.
```

Comment:
May we break down into more step? A long paragraph is a bit hard to understand.

Reply:
I'd suggest inline significant (pseudo) code to express a procedure; or at least a paragraph contain one link to one (kind of) method so that we get knowledge smoothly.

Reply:
```go
func (cc *clientConn) handleQuery(ctx context.Context, sql string) (err error) {
	stmts, err := cc.ctx.Parse(ctx, sql)
	// ...

	for i, stmt := range stmts {
		retryable, err = cc.handleStmt(ctx, stmt, parserWarns, i == len(stmts)-1)
		// ...
	}
}

func (cc *clientConn) handleStmt(ctx context.Context, stmt ast.StmtNode, warns []stmtctx.SQLWarn, lastStmt bool) (bool, error) {
	rs, err := cc.ctx.ExecuteStmt(ctx, stmt)
	// ...
}

func (tc *TiDBContext) ExecuteStmt(ctx context.Context, stmt ast.StmtNode) (ResultSet, error) {
	rs, err := tc.Session.ExecuteStmt(ctx, stmt)
	// ...
}

func (s *session) ExecuteStmt(ctx context.Context, stmtNode ast.StmtNode) (sqlexec.RecordSet, error) {
	// ...
	compiler := executor.Compiler{Ctx: s}
	stmt, err := compiler.Compile(ctx, stmtNode)
	// ...

	recordSet, err := runStmt(ctx, s, stmt)
	// ...
}

// Compile compiles an ast.StmtNode to a physical plan.
func (c *Compiler) Compile(ctx context.Context, stmtNode ast.StmtNode) (*ExecStmt, error) {
	// ...

	// PrepareTxnCtx starts a goroutine to begin a transaction if needed, and creates a new transaction context.
	s.PrepareTxnCtx(ctx)


	// Preprocess resolves table names of the node, and checks some statements validation.
	err := plannercore.Preprocess(c.Ctx, stmtNode, plannercore.WithPreprocessorReturn(ret), plannercore.WithExecuteInfoSchemaUpdate(pe))
	// ...

	// Optimize does optimization and creates a Plan.
	// The node must be prepared first.
	finalPlan, names, err := planner.Optimize(ctx, c.Ctx, stmtNode, ret.InfoSchema)
	
}
```

Is this helpful?

Reply:
Thanks, I added a slightly modified version of your proposal, PTAL.